src/solutions/ode_solutions.jl

```diff
@@ -425,12 +517,33 @@ function solution_new_tslocation(sol::ODESolution{T, N}, tslocation) where {T, N
         sol.original)
 end
 
+function solution_new_original_retcode(
+        sol::ODESolution{T, N}, original, retcode, resid) where {T, N}
+    return ODESolution{T, N}(
+        sol.u, sol.u_analytic, sol.errors, sol.t, sol.k, sol.discretes, sol.prob, sol.alg,
+        sol.interp, sol.dense, sol.tslocation, sol.stats, sol.alg_choice, retcode, resid,
+        original)
+end
+
 function solution_slice(sol::ODESolution{T, N}, I) where {T, N}
+    new_t = sol.t[I]
+    if sol.discretes === nothing
+        discretes = nothing
+    else
+        mint, maxt = extrema(new_t)
+        masked_discs = map(sol.discretes) do disc
+            mini = searchsortedfirst(disc.t, mint)
+            maxi = searchsortedlast(disc.t, maxt)
+            disc[mini:maxi]
```

Comment:
this implementation is only correct if `I` is a unit range? Should the argument type of `I` be restricted so that this method cannot be called with, say, `I = [1, 2, 4]`?

Reply:
Fixed in https://github.com/SciML/SciMLBase.jl/pull/645/commits/bb65e2dd37fd793d86686dfd14c1452acff4aa5f